// vue3-bootstrap-paginate/dist/vue3-bootstrap-paginate.es.js
import { defineComponent as $, toRef as V, computed as d, openBlock as u, createElementBlock as g, createElementVNode as p, normalizeClass as k, unref as a, withModifiers as h, toDisplayString as y, createCommentVNode as c, Fragment as J, renderList as E } from "vue";
const _ = { "aria-label": "..." }, A = {
  class: "page-link",
  href: "#"
}, F = /* @__PURE__ */ p("a", {
  class: "page-link",
  href: "#"
}, "1", -1), M = [
  F
], T = {
  key: 3,
  class: "page-item"
}, q = /* @__PURE__ */ p("span", { class: "page-link" }, "...", -1), z = [
  q
], L = ["onClick"], R = {
  class: "page-link",
  href: "#"
}, j = {
  key: 4,
  class: "page-item"
}, D = /* @__PURE__ */ p("span", { class: "page-link" }, "...", -1), G = [
  D
], H = {
  class: "page-link",
  href: "#"
}, I = {
  class: "page-link",
  href: "#"
}, O = /* @__PURE__ */ $({
  __name: "Vue3BsPaginate",
  props: {
    total: {
      type: Number,
      required: !0
    },
    perPage: {
      type: Number,
      default: 10,
      validator: (t) => {
        if (t <= 0) {
          const s = "'perPage' attribute must be greater than 0.";
          throw console.error(s), new TypeError(s);
        }
        return !0;
      }
    },
    pagesShown: {
      type: Number,
      default: 5,
      validator: (t) => {
        const s = "'pagesShown' attribute must be greater than 0.";
        if (t <= 0)
          throw console.error(s), new TypeError(s);
        return !0;
      }
    },
    modelValue: {
      type: Number,
      required: !0,
      validator: (t) => {
        const s = "'v-model' is required and must be greater than 0.";
        if (t <= 0)
          throw console.error(s), new TypeError(s);
        return !0;
      }
    },
    version: {
      type: Number,
      default: 5,
      validator: (t) => {
        const s = "bootstrap version 3, 4, 5";
        if (t <= 3 || t > 5)
          throw console.error(s), new TypeError(s);
        return !0;
      }
    },
    size: {
      type: String,
      default: "md"
    },
    alignment: {
      type: String,
      default: "start"
    },
    prevContent: {
      type: String,
      default: "\u2039"
    },
    nextContent: {
      type: String,
      default: "\u203A"
    },
    firstJumpContent: {
      type: String,
      default: "\xAB"
    },
    lastJumpContent: {
      type: String,
      default: "\xBB"
    },
    showPrevNextButton: {
      type: Boolean,
      default: !0
    },
    showJumpButtons: {
      type: Boolean,
      default: !1
    },
    onChange: {
      type: Function,
      default: () => {
      }
    }
  },
  emits: ["update:modelValue"],
  setup(t, { emit: s }) {
    const r = t, l = V(r, "modelValue"), o = d(() => Math.ceil(r.total / r.perPage)), m = d(() => {
      let n, e;
      if (o.value <= r.pagesShown)
        n = 1, e = o.value;
      else {
        let f = Math.floor(r.pagesShown / 2), w = Math.ceil(r.pagesShown / 2) - 1;
        l.value <= f ? (n = 1, e = r.pagesShown) : l.value + w >= o.value ? (n = o.value - r.pagesShown + 1, e = o.value) : (n = l.value - f, e = l.value + w);
      }
      let i = Array.from(Array(e + 1 - n).keys()).map((f) => n + f);
      return {
        total: r.total,
        currentPage: l.value,
        itemsPerPage: r.perPage,
        totalPages: o,
        startPage: n,
        endPage: e,
        pages: i
      };
    }), P = d(() => l.value === 1), C = d(() => l.value === o.value), B = d(() => m.value.pages[0] >= 2), b = d(() => m.value.pages[m.value.pages.length - 1] < o.value), S = d(() => m.value.pages[0] >= 4), x = d(() => m.value.pages[m.value.pages.length - 1] < o.value - 2), N = d(() => ["pagination", `pagination-${r.size}`, `justify-content-${r.alignment}`]), v = (n) => {
      n === l.value || n <= 0 || n > o.value || (s("update:modelValue", n), r.onChange(n));
    };
    return (n, e) => (u(), g("nav", _, [
      p("ul", {
        class: k(a(N))
      }, [
        t.showJumpButtons ? (u(), g("li", {
          key: 0,
          class: k(["page-item", { disabled: a(P) }])
        }, [
          p("a", {
            class: "page-link",
            href: "#",
            onClick: e[0] || (e[0] = h((i) => v(1), ["prevent"]))
          }, y(t.firstJumpContent), 1)
        ], 2)) : c("", !0),
        t.showPrevNextButton ? (u(), g("li", {
          key: 1,
          class: k(["page-item", { disabled: a(P) }]),
          onClick: e[1] || (e[1] = h((i) => v(a(l) - 1), ["prevent"]))
        }, [
          p("a", A, y(t.prevContent), 1)
        ], 2)) : c("", !0),
        a(B) ? (u(), g("li", {
          key: 2,
          class: "page-item",
          onClick: e[2] || (e[2] = h((i) => v(1), ["prevent"]))
        }, M)) : c("", !0),
        a(S) ? (u(), g("li", T, z)) : c("", !0),
        (u(!0), g(J, null, E(a(m).pages, (i, f) => (u(), g("li", {
          key: f,
          class: k(["page-item", { active: i === a(l) }]),
          onClick: h((w) => v(i), ["prevent"])
        }, [
          p("a", R, y(i), 1)
        ], 10, L))), 128)),
        a(x) ? (u(), g("li", j, G)) : c("", !0),
        a(b) ? (u(), g("li", {
          key: 5,
          class: "page-item",
          onClick: e[3] || (e[3] = h((i) => v(a(o)), ["prevent"]))
        }, [
          p("a", H, y(a(o)), 1)
        ])) : c("", !0),
        t.showPrevNextButton ? (u(), g("li", {
          key: 6,
          class: k(["page-item", { disabled: a(C) }]),
          onClick: e[4] || (e[4] = h((i) => v(a(l) + 1), ["prevent"]))
        }, [
          p("a", I, y(t.nextContent), 1)
        ], 2)) : c("", !0),
        t.showJumpButtons ? (u(), g("li", {
          key: 7,
          class: k(["page-item", { disabled: a(C) }])
        }, [
          p("a", {
            class: "page-link",
            href: "#",
            onClick: e[5] || (e[5] = h((i) => v(a(o)), ["prevent"]))
          }, y(t.lastJumpContent), 1)
        ], 2)) : c("", !0)
      ], 2)
    ]));
  }
});
export {
  O as Vue3BsPaginate
};
